Multi worker safety

ratulm opened the pull request:
Implemented queue management that can support multiple clients/workers. The code is complex now, and I need to figure out how to write unit tests for it. In the meanwhile, see if it passes the sniff test. Structurally, I tried to make it easy to follow. Happy to walk you through it.

This PR subsumes #731 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/741)
<!-- Reviewable:end -->

ratulm commented:
Figured out how to test and added some tests. Will add more later. If you can think of corner cases to worry about, I can convert those into tests.

corinaminer commented:
Reviewed 3 of 20 files at r1.
Review status: 15 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 1 of 20 files at r1, 1 of 2 files at r2, 1 of 1 files at r4.
Review status: 18 of 21 files reviewed at latest revision, 8 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 103 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GayJS8tuOt7DydbLn:-L0GayJS8tuOt7DydbLo:bshw6ny) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L103)):*
> ```Java
>                   envMetadata.getProcessingStatus()));
>         }
>         return currentParsingWork;
> ```

This line's unnecessary, it'll return `currentParsingWork` in the next case.

Same issue in `getBlockerForDataplaneIndependentWork`.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 198 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GuzHNcg2MXA1yflyY:-L0GuzHNcg2MXA1yflyZ:b-3thcl3) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L198)):*
> ```Java
>             "Unknown environment processingStatus: " + envMetadata.getProcessingStatus());
>     }
>   }
> ```

Should these `getBlocker` methods also be checking the delta environment if there is one?

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 238 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GwMMbQGbqxBhcQinu:-L0GwMMbQGbqxBhcQinv:b-bcvn68) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L238)):*
> ```Java
>   }
> 
>   private synchronized QueuedWork getIncompleteWorkType(
> ```

Could combine this with `getIncompleteWork` and make the `wType` parameter optional.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 441 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0Gy8bHT8XTFw5PmIll:-L0Gy8bHT8XTFw5PmIlm:bwrommi) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L441)):*
> ```Java
> 
>           // check if we unblocked anything
>           if (_blockingWork.contains(wItem.getId())) {
> ```

By the end of this if statement, the completed item should get removed from `_blockingWork`.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 463 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0Gzi3Csv8gD4KZyw81:-L0Gzi3Csv8gD4KZyw82:b-5kzlgs) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L463)):*
> ```Java
>                 // put this work on the completed queue in case people are checking its status
>                 _queueCompletedWork.enque(requeueWork);
>                 requeueWork.setStatus(WorkStatusCode.TERMINATEDABNORMALLY);
> ```

If this happens, should we run `processTaskCheckResult` on this work item?

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 558 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0H6L49kfTZjibHIE5Q:-L0H6L49kfTZjibHIE5R:byhrc7u) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L558)):*
> ```Java
>       return _queueIncompleteWork.enque(work);
>     }
>   }
> ```

Could simplify the above section by returning once we know it's not blocked:
```
if base blocked {
    ...
} else if differential {
    if delta blocked {
        ...
    }
}
return _queueIncompleteWork.enque(work);
```
Same for `queueDataplaneIndependentWork`.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 580 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0HDSBzHIzuDqVoDDFw:-L0HDSBzHIzuDqVoDDFx:btfml0q) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L580)):*
> ```Java
>       return _queueIncompleteWork.enque(work);
>     }
>   }
> ```

Could combine this and `queueDataplaneDependentWork` into one function with a boolean `dataplaneDependent` parameter

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 624 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0HFa1M94_f47j3DrWI:-L0HFa1M94_f47j3DrWJ:b-lohfy3) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L624)):*
> ```Java
>     // parsing work cannot proceeed in parallel because it may overwrite files used by others.
>     // instead of rejecting, we could have queued it as BLOCKED but we risk cycles of BLOCKED work
>     // this should not be a common case anyway, so we aren't losing much by rejecting it
> ```

Seems like this could make it difficult to re-parse a testrig (or regenerate its dataplane, in `queueDataplaningWork`) if there's a high volume of question requests. I agree that it's unlikely to be a concern at this point and not worth risking cycles of blocked work, but we should look for a better solution for the long term.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

ratulm commented:
Review status: 18 of 21 files reviewed at latest revision, 8 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 103 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GayJS8tuOt7DydbLn:-L0IAgnO2TCqJ3Q-xlHE:b-ec80rn) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L103)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

This line's unnecessary, it'll return `currentParsingWork` in the next case.

Same issue in `getBlockerForDataplaneIndependentWork`.
</blockquote></details>

oh i hate overflow across case statements. very risky move. 

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 198 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GuzHNcg2MXA1yflyY:-L0IMhDS6GeuW8aD_zil:b-7pt21t) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L198)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Should these `getBlocker` methods also be checking the delta environment if there is one?
</blockquote></details>

Good question. I think that is already taken care of however. Only PARSING and DATAPLANING works can be blockers, and these works are not differential. The works they block can be differential, but we are calling those getBlocking functions twice, once with respect to base and once with respect to delta. Blockers for dataplaning work need to be called only once since dataplaning work cannot be differential. With this context, you should be able to confirm that the right thing is happening. 

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 238 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0GwMMbQGbqxBhcQinu:-L0IOrTe48K1iPXhK4Ko:b-7rl0jd) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L238)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Could combine this with `getIncompleteWork` and make the `wType` parameter optional.
</blockquote></details>

done.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 441 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0Gy8bHT8XTFw5PmIll:-L0IPILQdQBKNHRCxPRA:b-kl8g4c) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L441)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

By the end of this if statement, the completed item should get removed from `_blockingWork`.
</blockquote></details>

good catch!

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 463 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0Gzi3Csv8gD4KZyw81:-L0IPUsZRqLN_VJ0Ztq9:b-jhj7mr) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L463)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

If this happens, should we run `processTaskCheckResult` on this work item?
</blockquote></details>

I thought about that but wasn't sure what the right approach should be. But reconsidering, that seems like the right thing to do.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 558 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0H6L49kfTZjibHIE5Q:-L0IVFDTMUKqiSlcMFAH:b-7rl0jd) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L558)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Could simplify the above section by returning once we know it's not blocked:
```
if base blocked {
    ...
} else if differential {
    if delta blocked {
        ...
    }
}
return _queueIncompleteWork.enque(work);
```
Same for `queueDataplaneIndependentWork`.
</blockquote></details>

done.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 580 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0HDSBzHIzuDqVoDDFw:-L0IWKew5oOet5_VzSXc:b-kghtrs) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L580)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Could combine this and `queueDataplaneDependentWork` into one function with a boolean `dataplaneDependent` parameter
</blockquote></details>

good point. done.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java, line 624 at r4](https://reviewable.io:443/reviews/batfish/batfish/741#-L0HFa1M94_f47j3DrWI:-L0IWRF-cYVitoPkkTKS:b-dnjnzy) ([raw file](https://github.com/batfish/batfish/blob/074d3bb642b66da81ec525eb9fb7d9b7a6e8f90a/projects/coordinator/src/main/java/org/batfish/coordinator/WorkQueueMgr.java#L624)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Seems like this could make it difficult to re-parse a testrig (or regenerate its dataplane, in `queueDataplaningWork`) if there's a high volume of question requests. I agree that it's unlikely to be a concern at this point and not worth risking cycles of blocked work, but we should look for a better solution for the long term.
</blockquote></details>

re-parsing and re-dataplaning are not going to be common activities hopefully, and limited to code debugging scenarios (where there may not be competition from answering). if it proves otherwise, yes, we'll need a different solution. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 3 of 6 files at r5.
Review status: 18 of 21 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 1 of 6 files at r5.
Review status: 19 of 21 files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 378 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R10zDEGOImljscrAX:-L0R10zDEGOImljscrAY:b6ozbt5) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L378)):*
> ```Java
>       Path qFile = getpathContainerQuestion(workItem.getContainerName(), qName);
>       Question question = Question.parseQuestion(qFile, getCurrentClassLoader());
>       isDifferential = question.getDifferential();
> ```

I think we should get `differential` from the work item, not the question file, because a non-differential question can be asked differentially by setting `ARG_DIFFERENTIAL` in the work item.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 399 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R4R-VTo4CVZqk5Rgr:-L0R4R-VTo4CVZqk5Rgs:b-3qihr4) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L399)):*
> ```Java
>         Question question = Question.parseQuestion(qFile, getCurrentClassLoader());
>         if (question.getDifferential()) {
>           isDifferential = true;
> ```

Should it be possible for an analysis to run some questions differentially and some non-differentially?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 378 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R10zDEGOImljscrAX:-L0RECpRCVanM9n1Mwra:b5dtxtt) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L378)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

I think we should get `differential` from the work item, not the question file, because a non-differential question can be asked differentially by setting `ARG_DIFFERENTIAL` in the work item.
</blockquote></details>

In the web UI, we set `DIFF_ACTIVE` whenever we want the question to be run using both base and delta testrigs.

We do not set `question.differential` unless the question is inherently differential (aka `reducedReachability`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 1 of 6 files at r5.
Review status: 20 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 137 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RBmYCb2xA7jNV7V5y:-L0RBmYCb2xA7jNV7V5z:bt6m2z4) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L137)):*
> ```Java
> 
>   @Test
>   public void qetMatchingWorkAbsent() throws Exception {
> ```

...is that "get" spelled with a Q?

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 149 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RECjaKR33MgiCvMkW:-L0RECjaKR33MgiCvMkX:b-ti63ep) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L149)):*
> ```Java
>     wItem3.addRequestParam("key3", "value3");
> 
>     QueuedWork matchingWork = _workQueueMgr.getMatchingWork(wItem3, QueueType.INCOMPLETE);
> ```

Could also do this before queueing anything, to make sure we get back `null` from an empty queue.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 155 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RESNWDusXBSwfvogB:-L0RESNWDusXBSwfvogC:bkdnr18) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L155)):*
> ```Java
> 
>   @Test
>   public void qetMatchingWorkPresent() throws Exception {
> ```

another non-standard spelling of get.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 174 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RFTB2eL7iVXkpFoGx:-L0RFTB2eL7iVXkpFoGy:bbb473q) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L174)):*
> ```Java
>   @Test
>   public void answeringIsRejectedForNonParsedBaseTestrigs() throws Exception {
>     initTestrigMetadata("testrig1", "env_default", ProcessingStatus.PARSING_FAIL);
> ```

any point in repeating this with `UNINITIALIZED` and `PARSING`?

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 178 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RFlDqsEa0moJIohfi:-L0RFlDqsEa0moJIohfj:bmx9m35) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L178)):*
> ```Java
>         new QueuedWork(
>             new WorkItem(CONTAINER, "testrig1"),
>             new WorkDetails("testrig1", "env_default", WorkType.DATAPLANE_DEPENDENT_ANSWERING));
> ```

same thing should happen for dataplane independent work if parsing isn't done, right? we should test that too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

ratulm commented:
Review status: 20 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 399 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R4R-VTo4CVZqk5Rgr:-L0RYvc5nnBYYVxQSGm0:bee1wzw) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L399)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Should it be possible for an analysis to run some questions differentially and some non-differentially?
</blockquote></details>

I was assuming that might happen. I wanted the code to be robust to that. (And its cheap to do.) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

ratulm commented:
Review status: 20 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 378 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R10zDEGOImljscrAX:-L0RZEzSHgIO2UZ874AV:bz6yw0) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L378)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

In the web UI, we set `DIFF_ACTIVE` whenever we want the question to be run using both base and delta testrigs.

We do not set `question.differential` unless the question is inherently differential (aka `reducedReachability`).
</blockquote></details>

Resolved in your conversation with Dan?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: 20 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 378 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R10zDEGOImljscrAX:-L0RZjZgEaltWXIsYJ5C:b-t4csv7) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L378)):*
> DIFF_ACTIVE

```
    if (_settings.getDifferential()) {
      question.setDifferential(true);
    }
    boolean dp = question.getDataPlane();
    boolean diff = question.getDifferential();
    boolean diffActive = _settings.getDiffActive() && !diff;
    _settings.setDiffActive(diffActive);
    _settings.setDiffQuestion(diff);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
What are all those lines removed from `Batfish.java`? Did they come from one of the merges with master?

---

Review status: 17 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 378 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R10zDEGOImljscrAX:-L0RoBiXNhUlQ0OQPOp6:bev37t2) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L378)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> DIFF_ACTIVE

```
    if (_settings.getDifferential()) {
      question.setDifferential(true);
    }
    boolean dp = question.getDataPlane();
    boolean diff = question.getDifferential();
    boolean diffActive = _settings.getDiffActive() && !diff;
    _settings.setDiffActive(diffActive);
    _settings.setDiffQuestion(diff);
```
</blockquote></details>

Resolution seems to be:
- Don't mess with `DIFF_ACTIVE`
- Attach a different marker (probably `ARG_DIFFERENTIAL`, which already exists) to work items that should be run differentially
- In `WorkMgr`, check `ARG_DIFFERENTIAL` instead of the question file to determine `isDifferential`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741#-:-L0Rp5X4WMGAL11H47CK:bhoydy5)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 1 of 20 files at r1, 2 of 4 files at r6.
Review status: 19 of 21 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 399 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R4R-VTo4CVZqk5Rgr:-L0RpsOPFPJZpdDNAIwO:b-733hq5) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L399)):*
<details><summary><i>Previously, ratulm wrote…</i></summary><blockquote>

I was assuming that might happen. I wanted the code to be robust to that. (And its cheap to do.) 
</blockquote></details>

But with this setup, it's not robust to that, right? As it stands, if any of the questions are differential, the whole analysis will get run differentially.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

ratulm commented:
Review status: 17 of 21 files reviewed at latest revision, 6 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 399 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0R4R-VTo4CVZqk5Rgr:-L0VzI7slFFxdbT8oYYS:b5huo9e) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L399)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

But with this setup, it's not robust to that, right? As it stands, if any of the questions are differential, the whole analysis will get run differentially.
</blockquote></details>

I think the right behavior is: as long as there is one differential question, we should declare the analysis as differential, so that the analysis gets appropriately blocked.

(Once this change is in, we can actually start queuing individual questions separately.)

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 137 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RBmYCb2xA7jNV7V5y:-L0RczQHxUT-Ui_ZGBg4:bgafr4d) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L137)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

...is that "get" spelled with a Q?
</blockquote></details>

of course :) (fixed)

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 149 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RECjaKR33MgiCvMkW:-L0RfhXqH971Q_scKIwX:b-7rl0jd) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L149)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

Could also do this before queueing anything, to make sure we get back `null` from an empty queue.
</blockquote></details>

done.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 174 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RFTB2eL7iVXkpFoGx:-L0WrxwwJ1d6TaFe-oR5:b-q3ko37) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L174)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

any point in repeating this with `UNINITIALIZED` and `PARSING`?
</blockquote></details>

added more tests. perhaps too many ...

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 178 at r5](https://reviewable.io:443/reviews/batfish/batfish/741#-L0RFlDqsEa0moJIohfi:-L0Ws1-gXNi3nLGFm6aq:b-q3ko37) ([raw file](https://github.com/batfish/batfish/blob/8c488242c4f7da7d9aa5c3ff54fc8a33857e30bc/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L178)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

same thing should happen for dataplane independent work if parsing isn't done, right? we should test that too.
</blockquote></details>

added more tests. perhaps too many ...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewable says there have been changes to `WorkMgr` since r5, but it's not showing me any changes. Should I just mark it as reviewed?

---

Reviewed 1 of 4 files at r6, 2 of 3 files at r7.
Review status: 20 of 21 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 36 at r7](https://reviewable.io:443/reviews/batfish/batfish/741#-L0f2GKfUZcZ5IZfyRO6:-L0f2GKfUZcZ5IZfyRO7:b-iqdko8) ([raw file](https://github.com/batfish/batfish/blob/faf0d455ad44c45e8cce99c48bfe83c91fa741e3/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L36)):*
> ```Java
>   private static final String BASE_ENV = "baseEnv";
> 
>   private static final String DELTA_TESTIG = "deltaTestrig";
> ```

`BASE_TESTRIG` and `DELTA_TESTRIG` are missing their Rs

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 593 at r7](https://reviewable.io:443/reviews/batfish/batfish/741#-L0f0eYuDVMkTMpFuQJ2:-L0f0eYuDVMkTMpFuQJ3:b-v3004k) ([raw file](https://github.com/batfish/batfish/blob/faf0d455ad44c45e8cce99c48bfe83c91fa741e3/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L593)):*
> ```Java
> 
>   @Test
>   public void ddAnsweringWithBaseParsingQueuedDataplaining() throws Exception {
> ```

"dataplaining": complaining with data to back you up

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741#-:-L0f4u6eDM1GmeFf4ZjO:bf3vg5u)*
<!-- Sent from Reviewable.io -->

ratulm commented:
There were a couple of changes, which are visible on github: https://github.com/batfish/batfish/commit/03b0e9580dee1fdef91fdffb6087bc0d9053459d

One minor bug fix, and one sanity check.

---

Review status: 20 of 21 files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 36 at r7](https://reviewable.io:443/reviews/batfish/batfish/741#-L0f2GKfUZcZ5IZfyRO6:-L0f_axdaP7uZ6Lj0yC2:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/faf0d455ad44c45e8cce99c48bfe83c91fa741e3/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L36)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

`BASE_TESTRIG` and `DELTA_TESTRIG` are missing their Rs
</blockquote></details>

Done.

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java, line 593 at r7](https://reviewable.io:443/reviews/batfish/batfish/741#-L0f0eYuDVMkTMpFuQJ2:-L0f_jeWHg1r1bkdeBO9:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/faf0d455ad44c45e8cce99c48bfe83c91fa741e3/projects/coordinator/src/test/java/org/batfish/coordinator/WorkQueueMgrTest.java#L593)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

"dataplaining": complaining with data to back you up
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741#-:-L0f_xcl_7DlQBvtYRxy:bw356ve)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>!

---

Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/741#-:-L0g6daTZTYLAu9BTX5O:bg7igkw)*
<!-- Sent from Reviewable.io -->
